src/pymor/algorithms/rand_la.py

```diff
@@ -2,21 +2,266 @@
+        assert lambda_min is None or isinstance(lambda_min, Number)
+        assert isinstance(complex, bool)
+        assert isinstance(self_adjoint, bool)
+
+        self.__auto_init(locals())
+        self._Q = [self.A.range.empty()]
+        for _ in range(subspace_iterations):
+            self._Q.append(self.A.source.empty())
+            self._Q.append(self.A.range.empty())
+        self._Q = tuple(self._Q)
+        self._samplevecs = self.A.range.empty()
+        self._adjoint_op = A if self_adjoint else AdjointOperator(A, range_product=range_product,
+                                                                  source_product=source_product)
+
+    @cached
+    def _lambda_min(self):
+        if isinstance(self.source_product, IdentityOperator):
+            return 1
+        elif self.lambda_min is None:
+            with self.logger.block('Estimating minimum singular value of source_product ...'):
+                return randomized_ghep(InverseOperator(self.source_product), n=1)[0]
+        else:
+            return self.lambda_min
+
+    def _draw_test_vector(self, n):
+        W = self.A.source.random(n, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(n, distribution='normal')
+        self._samplevecs.append(self.A.apply(W))
+
+    def _estimate_error(self, basis_size, num_testvecs, p_fail):
+        c = np.sqrt(2 * self._lambda_min())
+        c *= erfinv((p_fail / min(self.A.source.dim, self.A.range.dim)) ** (1 / num_testvecs))
+
+        if len(self._samplevecs) < num_testvecs:
+            self._draw_test_vector(num_testvecs - len(self._samplevecs))
+
+        W, Q = self._samplevecs[:num_testvecs].copy(), self._find_range(basis_size)
+        W -= Q.lincomb(Q.inner(W, self.range_product).T)
+        return c * np.max(W.norm(self.range_product))
+
+    def estimate_error(self, basis_size, num_testvecs=20, p_fail=1e-14):
+        r"""Randomized a posteriori error estimator for a given basis size.
+
+        This implements the a posteriori error estimator from :cite:`BS18` (Definition 3.1).
+
+        The error is given by
+
+        .. math::
+            \epsilon_{\mathrm{est}}=c_{\mathrm{est}}\cdot\max_{\omega\in\Omega}
+            \lVert (I-QQ^T)A\omega\rVert_{T\rightarrow S}
+
+        with
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\left(\frac{\texttt{p_fail}}{n}\right)^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`S` denotes the inner
+        product of the range, :math:`T` denotes inner product of the source of :math:`A` and
+        :math:`n=\min\{\dim(S),\,\dim(T)\}`.
+
+        Parameters
+        ----------
+        basis_size
+            The size of the basis.
+        num_testvecs
+            Number of test vectors for error estimation.
```

Comment:
I think some explanation should be given what enlarging the number of test vectors does, in particular in relation to `p_fail`.